ably/internal/ablyutil/time.go

```diff
@@ -57,16 +57,13 @@ func ContextWithTimeout(ctx context.Context, after TimerFunc, timeout time.Durat
 
 	// If the timer expires, we cancel the context. But then we need the context's
 	// error to be context.DeadlineExceeded instead of context.Canceled.
-	ctx, setErr := newContextWithCustomError(ctx)
+	// uses same context as above, doesn't create new context
+	ctx, setErr := contextWithCustomError(ctx)
 
 	go func() {
 		_, timerFired := <-after(ctx, timeout)
 		if timerFired {
 			setErr(context.DeadlineExceeded)
-		} else {
-			// after returned because the parent context was canceled, not
-			// because the timer fired.
-			setErr(ctx.Err())
```

Comment:
This was also necessary: if the timer didn't fire, then `ctx.Err()` is the underlying error that caused `after` to return, and it needs to be propagated.

Reply:
Yes, that's being taken care of here 
```func (ctx *contextWithCustomErr) Err() error {
	if ctx.err != nil {
		return ctx.err
	}
	return ctx.Context.Err()```

Reply:
Discussed offline.